connection-cache/src/connection_cache.rs

```diff
@@ -137,6 +138,12 @@ where
             .unwrap()
     }
 
+    pub fn update_key(&self, key: &Keypair) {
+        let mut map = self.map.write().unwrap();
+        map.clear();
+        let mut connection_manager = self.connection_manager.write().unwrap();
+        connection_manager.update_key(key);
```

Comment:
As far as I can see, the order is always self.map -> self.connection_manager so this should not cause a deadlock.

Reply:
Can we change the order of locking and release the lock so we don't hold the lock on both?
something like:
```rust
connection_manager.write().unwrap().update_key(key);
self.map.write().unwrap().clear();
```
This might discard some ok connections created between the 1st and the 2nd lock but I think that is better than the possibility of a dead-lock in the future.

Reply:
I just added a comment stating the required locking order. I think that is the better solution, since it ensures stricter correctness, and it's quite common to simply establish a locking order to prevent deadlocks, rather than not lock multiple locks at all.

Reply:
There is no way to enforce the comment.
Someone might change/add code later unaware of that comment there.
What is the problem with my suggestion above? https://github.com/solana-labs/solana/pull/33865#discussion_r1390311487

Reply:
> This might discard some ok connections created between the 1st and the 2nd lock

There's also no way to enforce that only 1 of the locks can be held at a time. It's not uncommon to just specify a locking order that future programmers must follow when dealing with multiple locks. Not a massive deal, but since we have 2 locks anyway, I would prefer to do the most strictly correct thing and clear the cache then update the key while holding the cache lock.